portfolio/src/main/java/com/google/sps/data/ReviewDaoDatastore.java

```diff
@@ -0,0 +1,204 @@
+package com.google.sps.data;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSet.Builder;
+import com.google.sps.data.Book;
+import com.google.sps.data.Review;
+import java.lang.Exception;
+import java.lang.NumberFormatException;
+import java.util.Set;
+
+public class ReviewDaoDatastore implements ReviewDao {
+  private DatastoreService datastore;
+  private static final String ENTITY_KIND = "Review";
+  private static final String ISBN_PROPERTY = "ISBN";
+  private static final String FULLTEXT_PROPERTY = "FullText";
+  private static final String USEREMAIL_PROPERTY = "UserEmail";
+  private static final String DEFAULT_EMAIL = "unknown@email.com";
+
+  public ReviewDaoDatastore() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void uploadAll(Book book) {
+    int userNumber = 0;
```

Comment:
What's the goal of this userNumber?

Reply:
This function is mainly for uploading the reviews from the CSV, at which point, we don't have user emails (which is why we're using the default email). 
But our ReviewDao by design doesn't allow the same user to leave multiple reviews for the same book. 
So without userNumber, we keep adding reviews by the same (default) user to the same book (an illegal action), which causes problems later when we're retrieving Reviews by a key (a Review key is for a unique email/isbn pair)

Reply:
It might be useful to document this in the code.